pkg/resourcemanager/webhook/nodeagentauthorizer/authorizer.go

```diff
@@ -199,6 +199,31 @@ func (a *authorizer) authorizePod(log logr.Logger, attrs auth.Attributes) (auth.
 		return auth.DecisionDeny, reason, nil
 	}
 
+	machine := &machinev1alpha1.Machine{}
+	if err := a.sourceClient.Get(ctx, client.ObjectKey{Name: machineName, Namespace: a.machineNamespace}, machine); err != nil {
+		return auth.DecisionDeny, "", fmt.Errorf("error getting machine %q: %w", machineName, err)
+	}
+
+	nodeName, ok := machine.Labels[machinev1alpha1.NodeLabelKey]
+	if !ok {
+		return auth.DecisionDeny, fmt.Sprintf(`"node" label not found on machine %q`, machineName), nil
+	}
+
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      attrs.GetName(),
+			Namespace: attrs.GetNamespace(),
+		},
+	}
+	if err := a.targetClient.Get(ctx, client.ObjectKeyFromObject(pod), pod); err != nil {
+		return auth.DecisionDeny, "", fmt.Errorf("error getting pod %q: %w", client.ObjectKeyFromObject(pod), err)
+	}
+
+	if pod.Spec.NodeName != nodeName {
+		log.Info("Denying request because pod belongs to a different node", "podName", attrs.GetName(), "podNamespace", attrs.GetNamespace(), "podNodeName", pod.Spec.NodeName)
+		return auth.DecisionDeny, fmt.Sprintf("pod %q does not belong to node %q", client.ObjectKeyFromObject(pod), nodeName), nil
+	}
+
```

Comment:
I don't think this will work for list and watch requests 😅 
You could try a similar thing [`node-authorizer`](https://github.com/kubernetes/kubernetes/blob/195803cde570ad1025a78e36cdbef76bddbc4c33/plugin/pkg/auth/authorizer/node/node_authorizer.go#L421-L435) is doing.